lib/cosmos/gui/text/ruby_editor.rb

```diff
@@ -196,19 +199,59 @@ def context_menu(point)
 
     def add_breakpoint(line)
       @breakpoints << line
+      block = document.findBlockByNumber(line-1)
+      block.setUserState(BREAKPOINT_SET)
+      block.dispose
+      block = nil
       @lineNumberArea.repaint
     end
 
     def clear_breakpoint(line)
       @breakpoints.delete(line)
+      block = document.findBlockByNumber(line-1)
+      block.setUserState(BREAKPOINT_CLEAR)
+      block.dispose
+      block = nil
       @lineNumberArea.repaint
     end
 
     def clear_breakpoints
       @breakpoints = []
+      block = document.firstBlock()
+      while (block.isValid())
+        block.setUserState(BREAKPOINT_CLEAR)
+        next_block = block.next()
+        block.dispose
+        block = next_block
+      end
+      block.dispose
+      block = nil
       @lineNumberArea.repaint
     end
 
+    def update_breakpoints
```

Comment:
Overall code looks great and works well.  I am worried about this method iterating over the whole document every time a line changes though.  For large documents this could be a killer.  Is there any way to do this only if breakpoints exist?